[bticinosmarther] Initial contribution (#7533)
openhab/openhab-addons: bundles/org.openhab.binding.bticinosmarther/README.md

# BTicino Smarther Binding

The Smarther binding implements a bridge to the Legrand/BTicino Smarther v2.0 API and allows you to control your BTicino Smarther chronothermostat units with openHAB, making it possible to discover Smarther units connected to your Legrand/BTicino Smarther account.

> Smarther chronothermostat is produced by BTicino (https://www.bticino.com/products-catalogue/smarther-the-connected-thermostat/), has its own API set and doesn't support the OpenWebNet protocol.

Reviewer: Avoid having long urls as plain text - better provide them as links.
```suggestion
Smarther chronothermostat is produced by [BTicino](https://www.bticino.com/products-catalogue/smarther-the-connected-thermostat/), has its own API set and does not support the OpenWebNet protocol.
```

> Companion HABPanel widget can be found here: [BTicino Smarther HABPanel widget](https://github.com/MrRonfo/openhab-smarther-widget)

Reviewer: There should not be specific references to a UI in this README, especially not from external sources.
```suggestion
```

> # Configuring the binding

Reviewer: Please structure the README along our [template](https://github.com/openhab/openhab-core/blob/master/tools/archetype/binding/src/main/resources/archetype-resources/README.md) and use the same headers, if possible - also make sure to capitalize all headers.

Author: Sure. In this case, what paragraph is best suitable to host the Bridge configuration process (for me to address the next suggestion too) ?

The binding requires you to register an Application with Legrand Developer portal at https://developer.legrand.com

> This will get you a set of Client ID and Client Secret parameters to be used by your binding configuration.

Reviewer: This is not really correct. The binding itself does not have ANY configuration in your case.
All you describe here is required for configuring a `BTicino Smarther Bridge`. Theoretically, you can have users with multiple applications registered at Legrand and using different bridges for them.

Author: Correct. See previous comment.

|-------------------|----------------------|------------|----------------------------------------------------------------------------------------|----------|
| apiCallsHandled   | Number               | Read-only  | The total number of API calls handled by the bridge                                    | Common   |
| notifsReceived    | Number               | Read-only  | The total number of push notifications received by the bridge                          | Common   |
| notifsRejected    | Number               | Read-only  | The total number of push notifications rejected by the bridge                          | Common   |

**Configuration Channels:**

The following channels represent convenience configuration channels for the bridge and must all be referenced with the `config#` prefix.

| Channel Type ID   | Item Type            | Read/Write | Description                                                                            | Type     |
|-------------------|----------------------|------------|----------------------------------------------------------------------------------------|----------|

> | fetchConfig       | Switch               | Read-write | Trigger to manually fetch updated bridge configuration from Smarther API               | Advanced |

Reviewer: This is not clear to me. What is the "bridge configuration" that is mentioned here? The bridge configuration should be set by the user, not by the external device.

Author: Maybe "bridge" configuration is a bit misleading. The bridge automatically gathers the remote device network topology from API. This is done on a periodical basis, as the API itself has a daily usage limit. The trigger is used to fetch the same information on demand. I can eventually change the description to "Trigger to manually fetch the updated remote device network topology"...

Reviewer: If manually triggering the fetch is a use case, I it ok for me to have such a channel (re-initializing the bridge would rather be a workaround for the user). So yes, if you could adapt the description to use vocabulary that does not collide with openHAB stuff, that would be nice!

> | accessToken       | String               | Read-only  | The current accessToken used in communication with Smarther API                        | Advanced |

Reviewer: accessTokens should never be a channel. What's the idea here?

Author: Not needed anymore, it was there since the beginning when I was exploring the OAuth2 framework.
Removed.

Reviewer: Yes, logging at trace or debug level is absolutely fine - those are not active by default, hence no security concerns here.

### Devices

> The temperature channels have dimension `Number:Temperature` and a precision of one tenth degree Celsius.

Reviewer: This is given in the table below - can this sentence here thus be removed?

Author: Is there just to give the detail on managed precision. Is it better to add it to field description (in the table) and remove the row ?

Reviewer: Yes, the info about dimension is duplicated here and the the one about precision is not given for other numbers in the table - so yes, I'd think it makes sense to add the info directly in the table for the respective channels.

| endMinute         | Number               | Read-write | New operational end minute to be set (valid only for "Manual" mode)                    | Common   |
| power             | Switch               | Read-write | Power on, send new operational settings to the module                                  | Common   |

_**Note:**_ The `program` and `endDate` channels are Selection channels.
They are dynamically populated by the binding, respectively with the module specific set programs and the next N days (starting from _"Today"_, N driven by the `numberOfEndDays` device configuration option).

**Configuration Channels:**

The following channels represent convenience configuration channels for the module and must all be referenced with the `config#` prefix.

| Channel Type ID   | Item Type            | Read/Write | Description                                                                            | Type     |
|-------------------|----------------------|------------|----------------------------------------------------------------------------------------|----------|

> | fetchConfig       | Switch               | Read-write | Trigger to manually fetch updated bridge configuration from Smarther API               | Advanced |

Reviewer: bridge configuration?

Author: Copy&Paste issue 😄

  }
}
```

bticinosmarther.rules:

```
rule "BTicino Smarther - Set time visibility flag"
when
    Item smaSetMode received update or Item smaSetDate received update
then
    if (smaSetMode.state == "MANUAL" && smaSetDate.state != "") {
        smaDisplayTime.sendCommand(ON)
    }
    else {
        smaDisplayTime.sendCommand(OFF)
    }
end
```

## Binding model and Smarther API

The model of the binding is such that the bridge takes care of all the remote communications with the Smarther API in the context of a specific user.
All devices (chronothermostats modules) currently associated with the user account are available to control.

Legrand/BTicino Smarther topology considers the following dimensions:

* topology : is the whole network of devices associated with a user account
* plant    : is the location where a module is installed (0..N plants per topology)
* module   : is the chronothermostat installed in a location (0..N modules per location)

You can add multiple bridges to allow controlling devices in the context of multiple Legrand user accounts.

Legrand manages the push notifications on device status via MS Azure C2C queues.

> Each bridge registers himself to the queue, takes care of incoming notifications on behalf of its managed devices and dispatches each payload to the related device.

Reviewer: ```suggestion
Each bridge registers itself to the queue, takes care of incoming notifications on behalf of its managed devices and dispatches each payload to the related device.
```

Author: LOL - fixed

> At the time of writing, the binding supports Smarther API v2.0, the last version released by Legrand.

Reviewer: ```suggestion
The binding uses the [Smarther API v2.0](https://portal.developer.legrand.com/docs/services/smartherV2/operations/Chronothermostat-Measures).

```

> The Smarther API and its documentation can be found at https://portal.developer.legrand.com/docs/services/smartherV2/operations/Chronothermostat-Measures.

Reviewer: ```suggestion
```
